SearchActivity.py

```diff
@@ -117,17 +146,57 @@ def _setup_toolbars(self):
             tooltip=_('Start a new game.'))
 
         self.status = label_factory(self.toolbar, '', width=300)
-
+        self.opponent = label_factory(self.toolbar, '')
         separator_factory(toolbox.toolbar, True, False)
 
         stop_button = StopButton(self)
         toolbox.toolbar.insert(stop_button, -1)
         stop_button.show()
 
+    # Collaboration-related methods
+
+    def _shared_cb(self, activity):
+        ''' Either set up initial share...'''
+        _logger.debug('shared')
+        self.after_share_join(True)
+
+    def _joined_cb(self, activity):
+        ''' ...or join an exisiting share. '''
+        _logger.debug('joined')
+        self.after_share_join(False)
+        self.send_nick()
+        # And let the sharer know we've joined
+        self.send_join()
+
+    def after_share_join(self, sharer):
+        self._game.set_sharing(True)
+        self._restoring = True
+
     def _new_game_cb(self, button=None):
         ''' Start a new game. '''
         self._game.new_game()
 
+    def _message_cb(self, collab, buddy, msg):
+        ''' Data from a tube has arrived. '''
```

Comment:
You can also remove or change this doc string as a text channel is used and not a tube.